File: library/src/main/java/cc/shinichi/library/tool/ui/PhoneUtil.java
package cc.shinichi.library.tool.ui;

import android.content.Context;
import android.util.DisplayMetrics;
import android.view.WindowManager;

/**
 * @author 工藤
 * @email 18883840501@163.com
 * cc.shinichi.drawlongpicturedemo.util
 * create at 2018/8/27  17:53
 * description:
 */
public class PhoneUtil {

    public static int getPhoneWid(Context context) {
        WindowManager windowManager =
                (WindowManager) context.getApplicationContext().getSystemService(Context.WINDOW_SERVICE);
        DisplayMetrics metric = new DisplayMetrics();
        windowManager.getDefaultDisplay().getMetrics(metric);
        return metric.widthPixels;
    }

    public static int getPhoneHei(Context context) {
        WindowManager windowManager =
                (WindowManager) context.getApplicationContext().getSystemService(Context.WINDOW_SERVICE);
        DisplayMetrics metric = new DisplayMetrics();
        windowManager.getDefaultDisplay().getMetrics(metric);
        return metric.heightPixels;
    }

    public static float getPhoneRatio(Context context) {
        return ((float) getPhoneHei(context)) / ((float) getPhoneWid(context));
    }

    public static int px2dp(Context context, float pxValue) {
        final float scale = context.getApplicationContext().getResources().getDisplayMetrics().density;
        return (int) (pxValue / scale + 0.5f);
    }

    public static int dp2px(Context context, float dipValue) {
        final float scale = context.getApplicationContext().getResources().getDisplayMetrics().density;
        return (int) (dipValue * scale + 0.5f);
    }
}

File: library/src/main/java/cc/shinichi/library/tool/ui/PhoneUtil.kt
package cc.shinichi.library.tool.ui

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.Resources
import android.os.Build
import android.util.DisplayMetrics
import android.util.Log
import android.view.Display
import android.view.WindowManager


/**
 * @author 工藤
 * @email qinglingou@gmail.com
 * cc.shinichi.drawlongpicturedemo.util
 * create at 2018/8/27  17:53
 * description:
 */
object PhoneUtil {

    private val TAG = PhoneUtil::class.java.simpleName

    fun getPhoneWid(context: Context): Int {
        val wm = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        val display: Display = wm.defaultDisplay
        var screenWidth = 0
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            val dm = DisplayMetrics()
            display.getRealMetrics(dm)
            screenWidth = dm.widthPixels
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.ICE_CREAM_SANDWICH) {
            screenWidth = try {
                Display::class.java.getMethod("getRawWidth").invoke(display) as Int
            } catch (e: Exception) {
                val dm = DisplayMetrics()
                display.getMetrics(dm)
                dm.widthPixels
            }
        }
        return screenWidth.apply {
            Log.d(TAG, "getPhoneWid: $this")
        }
    }

    fun getPhoneHei(context: Context): Int {
        val wm = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        val display = wm.defaultDisplay
        var screenHeight = 0

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            val dm = DisplayMetrics()
            display.getRealMetrics(dm)
            screenHeight = dm.heightPixels
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.ICE_CREAM_SANDWICH) {
            screenHeight = try {
                Display::class.java.getMethod("getRawHeight").invoke(display) as Int
            } catch (e: java.lang.Exception) {
                val dm = DisplayMetrics()
                display.getMetrics(dm)
                dm.heightPixels
            }
        }
        return screenHeight.apply {
            Log.d(TAG, "getPhoneHei: $this")
        }
    }

    fun getPhoneRatio(context: Context): Float {
        return (getPhoneHei(context).toFloat() / getPhoneWid(context).toFloat()).apply {
            Log.d(TAG, "getPhoneRatio: $this")
        }
    }

    @SuppressLint("InternalInsetResource")
    fun getNavBarHeight(context: Context): Int {
        val resources: Resources = context.resources
        val resourceId: Int = resources.getIdentifier("navigation_bar_height", "dimen", "android")
        return resources.getDimensionPixelSize(resourceId)
    }
}